src/orion/core/worker/multi_task_algo.py

```diff
@@ -0,0 +1,551 @@
+                    textwrap.dedent(
+                        f"""\
+                        Point is not contained in space:
+                        Point: {point}
+                        Space: {self.space}
+                        """
+                    )
+                )
+        return points_without_task_ids
+
+    def observe(
+        self, points: List[Point], results: Union[List[float], List[Dict]]
+    ) -> None:
+        """Observe evaluation `results` corresponding to list of `points` in
+        space.
+
+        .. seealso:: `orion.algo.base.BaseAlgorithm.observe`
+        """
+        # Add the task ids to the points, because we assume that the points passed to
+        # `observe` are from the current task.
+        if self._enabled:
+            # Add the task ids to the points.
+            points = [
+                self._add_task_id(point, self.current_task_id) for point in points
+            ]
+        return self.algorithm.observe(points, results)
+
+    def warm_start(self, warm_start_trials: Dict[Mapping, List[Trial]]) -> None:
+        """ Use the given trials to warm-start the algorithm.
+
+        These experiments and their trials were fetched from some knowledge base, and
+        are believed to be somewhat similar to the current on-going experiment.
+
+        It is the responsability of the Algorithm to implement this method in order to
+        take advantage of these points.
+
+        Parameters
+        ----------
+        warm_start_trials : Dict[Mapping, List[Trial]]
+            Dictionary mapping from ExperimentInfo objects (dataclasses containing the
+            experiment config) to the list of Trials associated with that experiment.
+        """
+        # Being asked to warm-start the algorithm.
+        if is_warmstarteable(self.algorithm):
+            log.debug(f"Warm starting the algo with trials {warm_start_trials}")
+            self.algorithm.warm_start(warm_start_trials)
+            log.info("Algorithm was successfully warm-started, returning.")
+            return
+
+        # Perform warm-starting using only the supported trials.
+        log.info(
+            "Will warm-start using contextual information, since the algo "
+            "isn't warm-starteable."
+        )
+
+        compatible_trials: List[Trial] = []
+        compatible_points: List[Tuple] = []
+        task_ids: list[int] = []
+
+        for i, (experiment_info, trials) in enumerate(warm_start_trials.items()):
+            # exp_space = experiment_info.space
+            # from warmstart.utils.api_config import hparam_class_from_orion_space_dict
+            # exp_hparam_class = hparam_class_from_orion_space_dict(exp_space)
+            log.debug(f"Experiment {experiment_info.name} has {len(trials)} trials.")
+            n_incompatible_trials = 0
+
+            for trial in trials:
+                try:
+                    # TODO: Use something smarter than `trial_to_tuple`, since it
```

Comment:
I think the filtering and adjustment (if needed) should be done outside of the algorithm. Filtering could be done by the knowledge base query, and adjustment could be done by something like the recycler you implemented for the warm-starting project or like the EVC adapters.